README.md
# just-auth

[![NPM][npm-badge-img]][npm-badge-link]
[![Build Status][travis-badge]][travis-badge-url]
[![Coverage Status][coveralls-badge]][coveralls-badge-url]

Simple SPA focused token based authentication for Express.js
This library follows convention over configuration, but configuration is available :wink:.

## Installation

Download node at [nodejs.org](http://nodejs.org) and install it, if you haven't already.

```sh
npm install just-auth --save
```


## Usage

```js
const express = require('express');
const justAuth = require('just-auth');
const app = express();
const auth = justAuth({
  secret: 'c47sRfunny101',

  getUser(email, callback) {
    // if error: callback({ myerror: 'failure' });
    // if success: callback(undefined, { email: 'my@email', passwordHash: '%asdaq42ad..' });
  },

  // Default behavior (don't specify if this suites you)
  configureToken(user) {
    // user without passwordHash
    return user;
  }
});

app.use('/auth', auth.router);

// Can also use `succeeded()` and `failed()` for redirects, etc.
// See https://www.npmjs.com/package/express-authentication
app.use('/api/admin', auth.middleware.required());

app.listen(80);
```

POST to `/auth/login` with `{ email: 'my@email', password: 'bacon' }`.
Result will be JSON, e.g. `{ token: '2mkql3...' }`.

_Note: To use the built in password utilities, you can use the following:_

```js
const passUtils = require('just-auth/lib/password');

const isValid = passUtils.validate(pass, hash);

passUtils.hash(pass, function (err, hash) {
  // error or hash
});
```


## Available Options

* `secret` - String, **required**.
* `loginEndpoint` - String, defaults to '/login'.
* `idField` - String, defaults to 'email', the field name of the identifier for the user.
  The value of this field is passed to the `getUser` function.
* `passwordField` - String, defaults to 'password'.
* `passwordHashField` - String, defaults to 'passwordHash'.
* `rememberMeField` - String, defaults to 'rememberMe'.
* `rememberMeAdditionalMinutes` - Number, defaults to 13 days in minutes.
* `tokenOptions` - Object, defaults to [this](https://github.com/knownasilya/just-auth/blob/master/index.js#L16). See full options
  [here](https://github.com/auth0/node-jsonwebtoken/tree/v7.0.0#jwtsignpayload-secretorprivatekey-options-callback).

### Methods

* `getUser` - **Required**; Function, `function (id, callback)`, should return a user object or an error via the callback.
* `configureToken` - Function, `function (user)`, should return the data that you want in the token, defaults to `user` if not specified.
* `validatePassword` - Function, `function (password, passwordHash)` should return a promise.
  By default this is `pbkdf2Utils.verify`, see [here](https://www.npmjs.com/package/pbkdf2-utils).


## Tests

```sh
npm install
npm test
````

## License

ISC

[npm-badge-img]: https://badge.fury.io/js/just-auth.svg
[npm-badge-link]: http://badge.fury.io/js/just-auth
[travis-badge]: https://travis-ci.org/knownasilya/just-auth.svg?branch=master
[travis-badge-url]: https://travis-ci.org/knownasilya/just-auth
[coveralls-badge]: https://coveralls.io/repos/knownasilya/just-auth/badge.svg?branch=master
[coveralls-badge-url]: https://coveralls.io/r/knownasilya/just-auth?branch=master


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [4.0.0](https://github.com/knownasilya/just-auth/compare/v3.0.0...v4.0.0) (2020-12-16)


### ⚠ BREAKING CHANGES

* Drop Node.js < 10
* when supplying `Bearer` (no token) you now get a 400 instead of a 401. See middleware - no token test.

### Bug Fixes

* update deps and fix tests ([242a7c5](https://github.com/knownasilya/just-auth/commit/242a7c5cb22a50a81aaeb3e2457c4d749ee22320))

<a name="3.0.0"></a>
# [3.0.0](https://github.com/knownasilya/just-auth/compare/v2.0.1...v3.0.0) (2018-04-16)


### Bug Fixes

* update deps and fix tests ([048ca47](https://github.com/knownasilya/just-auth/commit/048ca47))
* update the readme and removve readme script ([6f09bb0](https://github.com/knownasilya/just-auth/commit/6f09bb0))


### BREAKING CHANGES

* some errors are not the same in regards to the output and maybe the statuscode (400 vs 401).



## v2.0.1 (07/11/2016)

- Add more tests
- Fix invalid hash verification error returning

## v2.0.0 (07/11/2016)

- Dependencies updated
- Add Changelog

### Breaking Changes

- `tokenOptions` now takes on expiration option, `expiresIn`, which is in seconds.

### Features

- `var auth = justAuth({})` now returns `auth.createToken(user, rememberMe)` as a helper function for testing.


tests/index.js
'use strict';

const request = require('supertest');
const express = require('express');
const test = require('tape');
const boot = require('./bootstrap');
const helpers = require('./helpers');
const lib = require('../');
const hash =
  'IAQAAAfQKUkNkhzJPUtnQ0ZutmxXAdmsoDGGFU1xb4DZELh1Qo0PpHfiddHMOPGvceLuaKuH';

test('secret not set', function (t) {
  t.throws(
    function () {
      boot({});
    },
    /just-auth requires a `secret`/,
    'Missing secret throws error'
  );
  t.end();
});

test('invalid #getUser', function (t) {
  t.throws(
    function () {
      boot({
        secret: 'test',
      });
    },
    /getUser\(id, callback\)` function to be defined/,
    'Missing function throws descriptive error'
  );
  t.end();
});

test('exports correct object on init', function (t) {
  const instance = lib({
    secret: 'blah',
    getUser: function (cb) {
      cb();
    },
  });

  t.ok(instance.router && instance.router.length === 3, 'has express router');
  t.ok(
    instance.middleware && instance.middleware.required,
    'has express-authentication middleware object'
  );
  t.equal(typeof instance.createToken, 'function', 'has createToken exported');

  t.end();
});

test('login works', function (t) {
  const jwt = require('jsonwebtoken');
  const options = helpers.validBlankOptions({
    email: '<id>',
    passwordHash: hash,
  });

  request(boot(options))
    .post('/auth/login')
    .send({ email: 'blah@blah', password: 'bacon' })
    .expect('Content-Type', /json/)
    .expect(200)
    .end(function (err, res) {
      //console.log(res);
      t.error(err, 'No error');
      t.ok(res.body.token, 'Has token');

      const decoded = jwt.verify(res.body.token, options.secret);
      t.equal(decoded.email, 'blah@blah', 'Token valid');

      t.end();
    });
});

test('login works - rememberMe', function (t) {
  const jwt = require('jsonwebtoken');
  const options = helpers.validBlankOptions({
    email: '<id>',
    passwordHash: hash,
  });

  request(boot(options))
    .post('/auth/login')
    .send({
      email: 'blah@blah',
      password: 'bacon',
      rememberMe: true,
    })
    .expect('Content-Type', /json/)
    .expect(200)
    .end(function (err, res) {
      t.error(err, 'No error');
      t.ok(res.body.token, 'Has token');

      const decoded = jwt.verify(res.body.token, options.secret);
      const timestamp = Math.floor(Date.now() / 1000);
      const expectedExpiration = timestamp + 60 * 60 * 24 * 14;

      t.equal(decoded.email, 'blah@blah', 'Token valid');
      t.equal(decoded.exp, expectedExpiration, 'Expiration is correct');

      t.end();
    });
});

test('login wrong password', function (t) {
  const options = helpers.validBlankOptions({
    email: '<id>',
    passwordHash: hash,
  });

  request(boot(options))
    .post('/auth/login')
    .send({ email: 'blah@blah', password: 'wrongpass' })
    .expect('Content-Type', /json/)
    .expect(401)
    .end(function (err, res) {
      t.error(err);
      t.equal(res.body, 'Unauthorized');
      t.end();
    });
});

test('login invalid hash', function (t) {
  const options = helpers.validBlankOptions({
    email: '<id>',
    passwordHash: 'sdasd',
  });

  request(boot(options))
    .post('/auth/login')
    .send({ email: 'blah@blah', password: 'bacon' })
    .expect('Content-Type', /json/)
    .expect(400)
    .end(function (err, res) {
      t.error(err);
      t.equal(res.body, 'invalid hash length');
      t.end();
    });
});

test('login no hash', function (t) {
  const options = helpers.validBlankOptions({
    email: '<id>',
  });

  request(boot(options))
    .post('/auth/login')
    .send({ email: 'blah@blah', password: 'bacon' })
    .expect('Content-Type', /json/)
    .expect(400)
    .end(function (err, res) {
      t.error(err);
      t.equal(res.body, 'Invalid user data.');
      t.end();
    });
});

test('login getUser error passed on', function (t) {
  const options = helpers.validBlankOptions({
    email: false,
    passwordHash:
      '$2a$08$3hwGAN.NKAP/6VX3NdJ3zuDmEv0qfzXnOexwEzq2gT.rUk3ohx37y',
  });
  const instance = lib(options);
  const app = express();
  app.use('/auth', instance.router);

  request(app)
    .post('/auth/login')
    .send({ email: 'user', password: 'blah' })
    .expect('Content-Type', /json/)
    .expect(401)
    .end(function (err, res) {
      t.error(err, 'No error');
      t.same(res.body, { err: 'fail' }, 'Response passed on');
      t.end();
    });
});

test('invalid login body data', function (t) {
  request(boot(helpers.validBlankOptions()))
    .post('/auth/login')
    .send({ id: 'user', pass: 'blah' })
    .expect('Content-Type', /json/)
    .expect(400)
    .end(function (err, res) {
      t.error(err, 'No error');
      t.equal(
        res.body,
        'Invalid arguments, expected `email` and `password` to be present.',
        'Responds with required arguments'
      );
      t.end();
    });
});

test('middleware', function (t) {
  const options = helpers.validBlankOptions({
    email: '<id>',
    passwordHash: hash,
  });
  const instance = boot(options);
  const agent = request(instance);
  let token;

  agent
    .post('/auth/login')
    .send({ email: 'blah@blah', password: 'bacon' })
    .expect('Content-Type', /json/)
    .expect(200)
    .end(function (err, res) {
      t.error(err, 'No error for login');
      t.ok(res.body.token, 'Has token');

      token = res.body.token;

      agent
        .get('/admin')
        .set('Authorization', 'Bearer ' + token)
        .expect('Content-Type', /json/)
        .expect(200)
        .end(function (error, resp) {
          t.error(error, 'No error for restricted access');
          t.equal(resp.body, 'ok', 'Data accessed');
          t.end();
        });
    });
});

test('middleware - non bearer auth', function (t) {
  const options = helpers.validBlankOptions({
    email: '<id>',
    passwordHash: hash,
  });
  const instance = boot(options);
  const agent = request(instance);
  let token;

  agent
    .post('/auth/login')
    .send({ email: 'blah@blah', password: 'bacon' })
    .expect('Content-Type', /json/)
    .expect(200)
    .end(function (err, res) {
      t.error(err, 'No error for login');
      t.ok(res.body.token, 'Has token');

      token = res.body.token;

      agent
        .get('/admin')
        .set('Authorization', 'Tester ' + token)
        .expect('Content-Type', /json/)
        .expect(400)
        .end(function (error, resp) {
          t.error(error, 'No error for restricted access');
          t.equal(
            resp.body,
            "Invalid authorization scheme, expected 'Bearer'",
            'Invalid schema'
          );
          t.end();
        });
    });
});

test('middleware - invalid token', function (t) {
  const options = helpers.validBlankOptions({
    email: '<id>',
    passwordHash: hash,
  });
  const instance = boot(options);
  const agent = request(instance);
  let token;

  agent
    .post('/auth/login')
    .send({ email: 'blah@blah', password: 'bacon' })
    .expect('Content-Type', /json/)
    .expect(200)
    .end(function (err, res) {
      t.error(err, 'No error for login');
      t.ok(res.body.token, 'Has token');

      token = 'invalid';

      agent
        .get('/admin')
        .set('Authorization', 'Bearer ' + token)
        .expect('Content-Type', /json/)
        .expect(401)
        .end(function (error, resp) {
          t.error(error);
          t.same(resp.body, {
            status: 401,
            statusCode: 401,
            error: 'AUTHENTICATION_REQUIRED',
          });
          t.end();
        });
    });
});

test('middleware - malformed token', function (t) {
  const options = helpers.validBlankOptions({
    email: '<id>',
    passwordHash: hash,
  });
  const instance = boot(options);
  const agent = request(instance);

  agent
    .get('/admin')
    .set('Authorization', 'invalid')
    .expect('Content-Type', /json/)
    .expect(400)
    .end(function (error, resp) {
      t.error(error);
      t.same(resp.body, 'Invalid authorization supplied');
      t.end();
    });
});

test('middleware - no token', function (t) {
  const options = helpers.validBlankOptions({
    email: '<id>',
    passwordHash: hash,
  });
  const instance = boot(options);
  const agent = request(instance);

  agent
    .post('/auth/login')
    .send({ email: 'blah@blah', password: 'bacon' })
    .expect('Content-Type', /json/)
    .expect(200)
    .end(function (err, res) {
      t.error(err, 'No error for login');
      t.ok(res.body.token, 'Has token');

      agent
        .get('/admin')
        .set('Authorization', 'Bearer ')
        .expect('Content-Type', /json/)
        .expect(400)
        .end(function (error, resp) {
          t.error(error);
          t.same(resp.body, 'Invalid authorization supplied');
          t.end();
        });
    });
});

test('middleware - auth missing', function (t) {
  const options = helpers.validBlankOptions({
    email: '<id>',
    passwordHash: hash,
  });
  const instance = boot(options);
  const agent = request(instance);

  agent
    .post('/auth/login')
    .send({ email: 'blah@blah', password: 'bacon' })
    .expect('Content-Type', /json/)
    .expect(200)
    .end(function (err, res) {
      t.error(err, 'No error for login');
      t.ok(res.body.token, 'Has token');

      agent
        .get('/admin')
        .expect('Content-Type', /json/)
        .expect(400)
        .end(function (error, resp) {
          t.error(error);
          t.same(resp.body, 'Invalid authorization supplied');
          t.end();
        });
    });
});

test('password hash util', function (t) {
  const passwordUtil = require('../lib/password');

  passwordUtil.hash('abc', function (err, passHash) {
    t.error(err, 'No error');
    t.equal(typeof passHash, 'string', 'Hash is a string');
    t.ok(passwordUtil.validate('abc', passHash), 'Matches the password');
    t.end();
  });
});

test('createToken', function (t) {
  const instance = lib({
    secret: 'blah',
    getUser: function (cb) {
      cb();
    },
  });
  const token = instance.createToken({ blah: true }, false);

  t.ok(token, 'Token exists');
  t.end();
});


tests/helpers.js
'use strict';

exports.validBlankOptions = function (user) {
  user = user || {};

  return {
    secret: '123f3mmm34',
    getUser(id, callback) {
      if (user.email === '<id>') {
        user.email = id;
      }

      process.nextTick(function () {
        if (!user.email) {
          callback({ err: 'fail' });
        } else {
          callback(undefined, user);
        }
      });
    },
  };
};


tests/bootstrap.js
'use strict';

const express = require('express');
const morgan = require('morgan');
const lib = require('../');

module.exports = function (options) {
  const app = express();
  const instance = lib(options);

  app.use(morgan('dev'));
  app.use('/auth', instance.router);

  app.get('/admin', instance.middleware.required(), function (req, res) {
    res.json('ok');
  });

  app.use(function (err, req, res, next) {
    res.status(err.status || 500).json(err || req.authentication);
  });

  return app;
};


lib/login.js
'use strict';

const deleteKey = require('key-del');
const get = require('lodash.get');
const createToken = require('./create-token');

module.exports = function (options) {
  return function (req, res) {
    let getUser = options.getUser;
    let validatePassword =
      options.validatePassword || require('./password').validate;
    let body = req.body;
    let idField = options.idField;
    let passwordField = options.passwordField;
    let passwordHashField = options.passwordHashField;
    let rememberMeField = options.rememberMeField;
    let rememberMe = get(body, rememberMeField);
    let id = get(body, idField);
    let password = get(body, passwordField);

    if (id && password) {
      getUser.call({ req: req }, id, function (err, user) {
        if (err) {
          return res.status(401).json(err);
        }

        let hash;

        if (user && get(user, passwordHashField)) {
          hash = get(user, passwordHashField);

          validatePassword(password, hash)
            .then(function (valid) {
              let token;

              if (valid) {
                deleteKey(user, passwordHashField, { copy: false });
                token = createToken(user, rememberMe, options);

                res.json({ token: token });
              } else {
                res.status(401).json('Unauthorized');
              }
            })
            .catch(function (error) {
              res.status(400).json(error.message);
            });
        } else {
          res.status(400).json('Invalid user data.');
        }
      });
    } else {
      res
        .status(400)
        .json(
          'Invalid arguments, expected `' +
            idField +
            '` and `' +
            passwordField +
            '` to be present.'
        );
    }
  };
};


lib/middleware.js
'use strict';

const authHeader = require('auth-header');
const tokenUtils = require('./token');

module.exports = function (options) {
  return function (req, res, next) {
    let authorization = req.query.token
      ? 'Bearer ' + req.query.token
      : req.get('authorization');
    let auth;

    req.challenge = authorization;

    try {
      auth = authHeader.parse(authorization);
    } catch (error) {
      return res.status(400).json('Invalid authorization supplied');
    }

    if (auth && auth.scheme && auth.token) {
      if (auth.scheme !== 'Bearer') {
        return res
          .status(400)
          .json("Invalid authorization scheme, expected 'Bearer'");
      }

      tokenUtils.decode(auth.token, options, function (err, decoded) {
        if (err) {
          console.error(err);
          req.authenticated = false;
          req.authentication = err;
        } else {
          req.authenticated = true;
          req.authentication = decoded;
        }

        next();
      });
    } else {
      req.authenticated = false;
      req.authentication = 'Authorization token is missing';
      next();
    }
  };
};


index.js
'use strict';

const extend = require('extend');
const express = require('express');
const authentication = require('express-authentication')();
const bodyParser = require('body-parser');

let defaults = {
  loginEndpoint: '/login',
  idField: 'email',
  passwordField: 'password',
  passwordHashField: 'passwordHash',
  rememberMeField: 'rememberMe',
  // 13 days, totaling 2 weeks
  rememberMeAdditionalMinutes: 60 * 24 * 13,
  tokenOptions: {
    // A full day
    expiresIn: 60 * 60 * 24,
  },
};

module.exports = function (options) {
  const router = new express.Router();
  const loginHandler = require('./lib/login');
  const middleware = require('./lib/middleware');

  options = options || {};
  options = extend(true, {}, defaults, options);

  if (!options.secret) {
    throw new Error('just-auth requires a `secret` to be set in the options.');
  }

  if (!options.getUser || typeof options.getUser !== 'function') {
    throw new Error(
      'just-auth requires a `getUser(id, callback)` function to be defined. ' +
        'See https://github.com/knownasilya/just-auth#getuser for details.'
    );
  }

  // Setup router specific middleware
  router.use(bodyParser.json());
  router.use(
    bodyParser.urlencoded({
      extended: false,
    })
  );

  // Login endpoints
  router.route(options.loginEndpoint).post(loginHandler(options));

  const authMiddleware = authentication
    .for('just-auth')
    .use(middleware(options));

  return {
    router: router,
    middleware: authMiddleware,
    createToken: function createTokenExternal(data, remember) {
      const createToken = require('./lib/create-token');
      return createToken(data, remember, options);
    },
  };
};
